compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/Fir2IrVisitor.kt

```diff
@@ -485,7 +486,8 @@ class Fir2IrVisitor(
                 }
             }
         }
-        return implicitCastOrExpression(value, castTypeRef.toIrType())
+        return implicitCastOrExpression(value, castTypeRef.toIrType(),
+                                        expressionWithSmartcast.originalExpression is FirThisReceiverExpression)
```

Comment:
Why don't you check here that `this` expression is from local anonymous function?

Reply:
added checks for local functions